R/guide-legend.R

```diff
@@ -225,31 +225,36 @@ GuideLegend <- ggproto(
 
   get_layer_key = function(params, layers, data) {
 
+    # Return empty guides as-is
+    if (nrow(params$key) < 1) {
+      return(params)
+    }
+
     decor <- Map(layer = layers, df = data, f = function(layer, df) {
 
+      # Subset key to the column with aesthetic matching the layer
       matched_aes <- matched_aes(layer, params)
+      key <- params$key[matched_aes]
+      key$.id <- seq_len(nrow(key))
 
-      if (length(matched_aes) > 0) {
-        # Filter out aesthetics that can't be applied to the legend
-        n <- lengths(layer$aes_params, use.names = FALSE)
-        layer_params <- layer$aes_params[n == 1]
+      # Filter static aesthetics to those with single values
+      single_params <- lengths(layer$aes_params) == 1L
+      single_params <- layer$aes_params[single_params]
 
-        aesthetics  <- layer$computed_mapping
-        is_modified <- is_scaled_aes(aesthetics) | is_staged_aes(aesthetics)
-        modifiers   <- aesthetics[is_modified]
+      # Use layer to populate defaults
+      key <- layer$compute_geom_2(key, single_params)
 
-        data <- layer$geom$use_defaults(params$key[matched_aes], layer_params, modifiers)
-        data$.draw <- keep_key_data(params$key, df, matched_aes, layer$show.legend)
-      } else {
-        reps <- rep(1, nrow(params$key))
-        data <- layer$geom$use_defaults(NULL, layer$aes_params)[reps, ]
```

Comment:
The visual test for the non-sf legend is to cover this case, which gets solved by Layer$compute_geom_2(), but was never triggered in the test base.